# Cours 3 : Récursivité

## Rappel : la pile d'exécution

Au cours de l'exécution d'un programme les variables sont stockées dans la mémoire sous forme d'une **pile**. A chaque appel de fonction, un espace mémoire est réservé pour les variables de la fonction.

**Exemple** 

In [1]:
def fonction1(a):
    c = a + 1
    return c

def fonction2(b):
    return fonction1(b+1)

a = fonction2(3)
a

5

Voici une image simplifiée de l'évolution de la pile lors de l'exécution

![Evolution de la pile lors de l'exécution](pile1.png)

On va utiliser le petit bout de code suivant pour suivre en direct les appels de fonctions, exécutez la cellule qui définit la fonction `printAppelsFonctions` puis regardons comment on l'applique sur notre petit bout de code.

In [2]:
import functools

def printAppelsFonctions(func):
    """Print the function signature and return value"""
    @functools.wraps(func)
    def wrapper_debug(*args, **kwargs):
        args_repr = [repr(a) for a in args]                      
        kwargs_repr = [f"{k}={v!r}" for k, v in kwargs.items()]  
        signature = ", ".join(args_repr + kwargs_repr)           
        print(f"Appel de {func.__name__}({signature})")
        value = func(*args, **kwargs)
        print(f"valeur de retour {value!r}")
        print(f"Fin de {func.__name__}({signature})")           
        return value
    return wrapper_debug

In [3]:
@printAppelsFonctions
def fonction1(a):
    c = a + 1
    return c

@printAppelsFonctions
def fonction2(b):
    return fonction1(b+1)

a = fonction2(3)
a

Appel de fonction2(3)
Appel de fonction1(4)
valeur de retour 5
Fin de fonction1(4)
valeur de retour 5
Fin de fonction2(3)


5

**Exemple**

que se passe-t-il dans le cas suivant ?

In [4]:
def fonction1(a):
    fonction1(a)

fonction1(3)

RecursionError: maximum recursion depth exceeded

In [0]:
@printAppelsFonctions
def fonction1(a):
    fonction1(a)

fonction1(3)

## Définition

Un algorithme *récursif* est un algorithme qui fait appel à lui même. Lorsqu'un algorithme n'est pas récursif, on dit qu'il est *itératif*.

Pour éviter une boucle infinie, il faudra impérativement définir **une condition d'arrêt**.

**Exemple**

Le calcul de $n! = n \times (n-1) \times \dots \times 1$. Tout d'abord la version itérative

In [4]:
def factorielleIterative(n):
    r = 1
    for i in range(1,n+1):
        r = r*i
    return r

In [5]:
factorielleIterative(3)

6

In [6]:
factorielleIterative(5)

120

La version récursive se base sur le principe suivant :

$n! = \begin{cases}
1 & \text{si } n = 0 \\
n \times (n-1)! & \text{sinon}
\end{cases}$

In [7]:
def factorielleRecursive(n):
    if n == 0:
        return 1
    return n * factorielleRecursive(n-1)

In [8]:
factorielleRecursive(3)

6

In [9]:
factorielleRecursive(5)

120

Que se passe-t-il au niveau de la pile ?

In [10]:
@printAppelsFonctions
def factorielleRecursive(n):
    if n == 0:
        return 1
    return n * factorielleRecursive(n-1)


In [11]:
factorielleRecursive(3)

Appel de factorielleRecursive(3)
Appel de factorielleRecursive(2)
Appel de factorielleRecursive(1)
Appel de factorielleRecursive(0)
valeur de retour 1
Fin de factorielleRecursive(0)
valeur de retour 1
Fin de factorielleRecursive(1)
valeur de retour 2
Fin de factorielleRecursive(2)
valeur de retour 6
Fin de factorielleRecursive(3)


6

Attention, sur certaines valeurs, on obtient des appels récurisfs à l'infini. Ici, la *fonction termine* sur l'ensemble des entiers positifs mais pas sur les négatifs

In [0]:
factorielleRecursive(-2)

## Comment définir une fonction récursive ?

Le principe d'une fonction récursive est assez similaire à la notion de **récurrence** mathématique. On ne cherche pas à calculer le problème dans son ensemble mais simplement à le définir en fonction d'un problème de taille plus petite. Tout comme en mathématique le fait de prouver l'état initial et une seule étape permet de prouver la propriété, dans un algorithme, le fait de calculer l'état initial et une étape permet de calculer n'importe quelle valeur. 

Un algorithme récursif s'écrira toujours de cette façon :

Pour être sûr que l'algorithme termine, les nouveaux paramètres doivent se rapprocher de la condition d'arrêt.

## Exercices

### Exercice 1

1. Donner un algorithme récursif qui affiche les entiers de 1 à $n$
1. Même chose de $n$ à 1
3. Puis-je modifier ces algorithmes pour qu'ils rajoutent "fin" à la fin de la liste

### Exercice 2

Pour chacun des algorithmes suivants :

1. Déterminez pour quelles valeurs d'entrée l'algorithme termine (= ne fait pas de boucles infinies)
2. Calculez un exemple à la main
3. Expliquez ce que calcule l'algorithme

In [23]:
def fonction1(n):
    if n == 0:
        return 1
    return fonction1(n+1)

In [25]:
def fonction2(n):
    if n == 0:
        return 0
    return fonction2(n-1)+n

In [27]:
def fonction3(n):
    if n == 0:
        return 0
    return fonction3(n-1) - n

In [29]:
def fonction4(n):
    if n == 0:
        return 0
    if n < 0:
        return n + fonction4(-n)
    return n + fonction4(-n+1)

In [33]:
def fonction5(n):
    if n <= 1:
        return 0
    return 1 + fonction5(n-2)

### Exercice 3

La fonction de fibonacci est définie par :

$U_0 = 1$

$U_1 = 1$

$U_n = U_{n-1} + U_{n-2}$ si $n \geq 2$

Ces premières valeurs sont donc : 1, 1, 2, 3, 5, 8, 13, ...

Donner deux algorithmes, un itératif et un récursif, qui calculent la valeur $n$ de la suite.


### Exercice 4

On rappelle sur un exemple le principe de l'algorithme d'Euclide du calcul du pgcd (plus grand diviseur commun). 

Calcul du pgcd de 2145 et 630 : On commence par effectuer la division euclidienne de 2145 par 630

$2145 = 630 * 3 + 255$

puis on effectue la division de 630 par le reste obtenu 255

$630 = 255 * 2 + 120$

On continue jusqu'à ce que l'on trouve un reste nul.

$\begin{align*}
255 &= 120 *2 + 15 \\
120 &= 15 * 8
\end{align*}$

Le pgcd est le dernier reste non nul, c'est-à-dire **15**.

Donner un algorithme récursif qui calcule le pgcd de deux nombres.

### Exercice 5

Reprendre l'algorithme de recherche dichotomique dans un tableau trié vu dans le TD1 et donner une version récursive.

### Exemple avancé : les Tours de Hanoï

![Tour de Hanoï](hanoi.jpg)

Problème : on possède trois pics sur lesquels sont empilés des disques de tailles décroissantes (tous sur le premier pic). On souhaite déplacer la pile sur le second pic en suivant les règles suivantes:

* on ne peut déplacer qu'un seul disque à la fois, celui du haut de la pile,
* on ne peut poser un disque que sur un disque plus grand.


On suppose donc que l'on dispose d'une seule action possible :

En python, on utilisera l'interface suivante :

In [39]:
class Towers:
    
    def __init__(self,n):
        self.towers = (list(range(n,0,-1)),[],[])
        self.size = n
    
    def move(self,i,j):
        t1,t2 = self.towers[i], self.towers[j]
        if len(t1) == 0 or (len(t2) != 0 and t1[-1]>t2[-1]):
            raise ValueError("Invalid action")
        print("Move from Tower " + str(i) + " to Tower " + str(j))
        t2.append(t1.pop())
    
    def __repr__(self):
        res = ""
        n = self.size
        for i in range(n):
            line = ""
            for t in self.towers:
                if len(t) > i:
                    v = t[i]
                else:
                    v = 0
                stars = v*"*"
                spaces = (n-v)*" "
                line+= spaces + stars + "|" + stars + spaces + " "
            line+="\n"
            res = line + res
        return res

In [40]:
T = Towers(3)
T

  *|*      |       |    
 **|**     |       |    
***|***    |       |    

In [41]:
T.move(0,1)
T

Move from Tower 0 to Tower 1


   |       |       |    
 **|**     |       |    
***|***   *|*      |    

**Comment résoudre le problème ?**

Pour résoudre un problème de façon récursive, il faut répondre à plusieurs questions :

1. Quelle est la taille de mon problème ?
2. Quelle est la plus petite taille possible ? (condition d'arrêt)
3. Comment résoudre le problème dans ce cas là ? (action d'arrêt)
4. Si je suppose que je sais résoudre le problème pour toutes les tailles $k < n$, comment le résoudre pour la taille $n$ ?


## Complexité d'un algorithme récursif

### méthode de calcul

Reprenons le schéma de base d'un algorithme récursif  que l'on précise légèrement:

On veut compter le nombre d'actions de base. Chaque appel de fonction est une action de base. Soit $f$ la complexité de l'algorithme. Pour simplifier l'écriture, on suppose que les paramètres de l'algorithme correspondent à la taille $n$ du problème. La condition d'arrêt est obtenue quand $n=0$ et correspond à une complexité constante $c_0$. On obtient une définition récursive de $f$

$\begin{align}
f(0) &= c_0 \\
f(n) &= g(n) + \sum_{i=0}^k f(p_i)
\end{align}$

où $g$ est la complexité de la partie itérative. La complexité dépend donc du \textbf{nombre d'appels récursifs} à chaque étape et de la **taille des paramètres**. Il peut être plus ou moins simple de *développer* la fonction $f$ pour obtenir une formule close (non récursive).

### Cas classiques

On va regarder quelques cas classiques et compter le nombre d'appels de fonction. Pour cela, on utilise la fonctionnalité suivante (exécuter la cellule)

In [52]:
import functools

COMPTEUR = 0
STACK = 0

def compteAppels(func):
    c = 0
    """Print the function signature and return value"""
    @functools.wraps(func)
    def wrapper_debug(*args, **kwargs):
        global COMPTEUR, STACK
        COMPTEUR +=1
        STACK += 1
        value = func(*args, **kwargs)
        STACK -=1
        if STACK == 0:
            print(f"{COMPTEUR} appels à la fonction {func.__name__}")
            COMPTEUR = 0
        return value
    return wrapper_debug

**Exemple : cas linéaire**

On reprend la fonction factorielle vue précédemment.

In [53]:
@compteAppels
def factorielle(n):
    if n <= 0:
        return 1
    return n*factorielle(n-1)

In [55]:
factorielle(3)

4 appels à la fonction factorielle


6

Lorsqu'on appelle `factorielle(n)` on génère $n+1$ appels de fonctions. On est sur le schéma suivant :

Dans ce cas on a 

$\begin{align}
f(0) &= O(1) \\
f(n) &= 1 + f(n-1)
\end{align}$

Lorsqu'on développe, on obtient

$\begin{equation}
f(n) = 1 + 1 + 1 + \dots + 1 
\end{equation}$

$n+1$ fois.

Conclusion : la complexité est en $O(n)$.

**Exemple : cas exponentiel**

Reprenons l'exemple des Tours de Hanoï. Combien d'appels de fonctions sont réalisés pour la taille $n$ ?

In [49]:
@compteAppels
def Hanoi(n, T, depart, arrivee, inter):
    if n == 0:
        return
    Hanoi(n-1, T, depart, inter, arrivee)
    T.move(depart, arrivee)
    Hanoi(n-1, T, inter, arrivee, depart)

In [60]:
n = 1
T = Towers(n)
Hanoi(n,T,0,1,2)
T

Move from Tower 0 to Tower 1
3 appels à la fonction Hanoi


 |  *|*  |  

On est sur le schéma suivant :

Dans ce cas, la complexité est donnée par le nombre d'appels $f$ avec

$\begin{align}
f(0) &= 1 \\
f(n) &= 1 + 2 \times f(n-1).
\end{align}$

Par exemple, pour $n=3$ :

$\begin{align}
f(3) &= 1 + 2f(2) = 1 + 2(1 + 2f(1)) = 1 + 2(1 + 2(1 + 2))\\
 &= 15 = 2^4 - 1.
\end{align}$

On prouve facilement par récurrence que
$\begin{equation}
f(n) = 2^{n+1} - 1.
\end{equation}$

En effet, $f(0) = 1 = 2^1 - 1$ et si on suppose la formule vraie pour $n$, on obtient pour $n+1$

$f(n+1) = 1 + 2 f(n) = 1 + 2(2^{n+1} -1) = 1 + 2^{n+2} - 2 = 2^{n+2} - 1$

On est dans le cas d'une complexité exponentielle en $O(2^n)$.

**Exemple : cas logarithmique**

Reprenons l'exemple de la fonction de recherche dichotomique. Combien d'appels sont réalisés ici ?

In [63]:
@compteAppels
def rechercheDich(T,v,deb, fin):
    m = (deb+fin)//2
    if deb == fin:
        return False
    if T[m] == v:
        return True
    if v < T[m]:
        return rechercheDich(T,v,deb,m)
    return rechercheDich(T,v,m+1,fin)

In [78]:
n = 7
T = list(range(n))
v = 0.5
rechercheDich(T,v,0,n)

4 appels à la fonction rechercheDich


False

On est sur le schéma suivant

Dans ce cas, si $f$ est le nombre d'appels, on a

$\begin{align}
f(0) &= 1 \\
f(n) &= 1 + f(\lfloor n/2 \rfloor).
\end{align}$

Développons sur un exemple :

$\begin{align}
f(10) &= 1 + f(5) = 2 + f(2) = 3 + f(1) = 5 \\
&= \log(10) + 2.
\end{align}$

(on prend le $\log$ en base 2)

On prouve par récurrence
$\begin{equation}
f(n) = \lfloor \log(n) \rfloor + 2.
\end{equation}$

En effet, 

$f(1) = 1 + f(0) = 2 = \log(1) + 2$ 

En supposant la formule correcte pour toute valeur inférieure stricte à $n$, on obtient

$\begin{align}
f(n) &= 1 + f(\lfloor n/2 \rfloor) \\
&= 1 + \lfloor \log(\lfloor  n/2 \rfloor) \rfloor + 2
\end{align}$

Supposons $n$ pair, si $k = \lfloor \log(n/2) \rfloor$, cela signifie

$2^k \leq \frac{n}{2} < 2^{k+1}$, c'est à dire

$2^{k+1} \leq n < 2^{k+2}$

On peut obtenir une inégalité similaire dans le cas de $n$ impair. Cela signifie que :

$\lfloor \log(\lfloor  n/2 \rfloor) \rfloor = \lfloor \log(n) \rfloor - 1$ 

et donc

$f(n) = 1 + \lfloor \log(n) \rfloor - 1 + 2 = \lfloor \log(n) \rfloor + 2$

On a donc une complexité en $O(log(n))$.

### Conclusion : calcul de complexité

Si l'on simplifie le problème aux cas où la partie itérative de la fonction est en $O(1)$ (comme dans les exemples précédents). La complexité est donnée entièrement par le nombre d'appels récursifs, $f$ on a 3 cas de figures. ($v$ et $k$ sont des constantes qui ne dépendent pas de $n$)

$f(n) =  1 + f(n-v) \rightarrow$ complexité linéaire $O(n)$

$f(n) = 1 + k f(n - v) \rightarrow$ complexité exponentielle $O(k^n)$

$f(n) = 1 + f(n/k) \rightarrow$ complexité logarithmique $O(\log(n))$ (la constante $k$ agit sur la base dans laquelle on prend le logarithme, ce qui ne change pas la classe de complexité)

## Exercices

### Exercice 6

Dans la fonction suivante du calcul des nombres de Fibonacci, combien d'appels de fonctions sont réalisés pour un $n$ donné ? Quelle est la complexité ?

In [79]:
def fibo(n):
    if n <= 1:
        return 1
    return fibo(n-1) + fibo(n-2)

In [82]:
fibo(5)

8

### Exercice 7

Il existe deux définitions récursives de la fonction puissance :

$\begin{align*}
a^b &= \begin{cases}
1 &\text{si }b=0 \\
a \times a^{b-1} &\text{sinon}
\end{cases} \\
a^b &= \begin{cases}
1 &\text{si }b=0 \\
a \times a^{b-1} &\text{si }b\text{ est impair} \\
a^{\frac{b}{2}}a^{\frac{b}{2}} &\text{si }b\text{ est pair}
\end{cases}
\end{align*}$

Pour chacune des définitions, donner l'algorithme récursif correspondant ainsi que sa complexité.

## Pour finir : itératif ou récursif ?

En terme de complexité, la récursivité ne permet pas par défaut d'obtenir des algorithmes plus efficaces. Quand elle est mal utilisée, elle peut même donner des complexité **plus mauvaise** : exemple de Fibonacci. Par ailleurs, on peut prouver qu'il existe **toujours** une version itérative d'un algorithme récursif : il suffit de déplier la pile !

**Qu'en est-il de la complexité mémoire ?** Reprendre l'exemple de la fonction exponentielle en itératif et en récursif. La complexité mémoire de l'algorithme récursif est en $O(n)$ tandis qu'il est en $O(1)$ pour l'itératif.

**Alors pourquoi on fait du récursif ?** Sur de nombreux problèmes, les algorithmes récursifs sont beaucoup plus simples à concevoir que les algorithmes itératifs. C'est le cas de l'algorithme des Tour de Hanoi. Ils donnent des codes plus courts et plus lisibles. Par ailleurs, certaines **structure de données** ont elles-mêmes des **définition récursives** et sont particulièrement adaptés aux algorithmes récursifs : les arbres, les graphes, les listes chaînées. 